src/oge/anomaly_screening.py

```diff
@@ -0,0 +1,430 @@
+from typing import Any, Optional
+
+import numpy as np
+import pandas as pd
+
+from oge.logging_util import get_logger
+
+logger = get_logger(__name__)
+
+
+def flag_cems_outliers(cems, field, global_cut_multiplier=10):
+    """Return a data frame giving for each plant amd emission unit combination in CEMS
+    data the number of missing values, global extreme and the mean deviation of the
+    global extreme relative to the median value of the time series.
+
+    Arguments
+    ---------
+    * `cems` : CEMS data as a data frame.
+    * `field` : Column to screen for global extreme in CEMS data frame.
+    * `global_cut_multiplier` : Cut above which values in the time series will be
+                                considered as global extreme. Express as multiple of
+                                median value.
+    """
+    screening = []
+    for i, ts in enumerate(cems.groupby(["plant_id_eia", "emissions_unit_id_epa"])):
+        AS = AnomalyScreeningFirstStep(
+            ts[1][["plant_id_eia", "emissions_unit_id_epa", field]],
+            field,
+            global_cut_multiplier=global_cut_multiplier,
+        )
+        AS.flag_negative_values()
+        AS.flag_zero_values()
+        if (
+            len(
+                AS.get_filtered_df().query(
+                    "category != 'MISSING' and category != 'ZERO'"
+                )
+            )
+            > 0
+        ):
+            AS.flag_global_extreme_values()
+            df = AS.get_filtered_df()
+            if len(df.query("category == 'GLOBAL_EXTREME'")) > 0:
+                global_extreme_id = df.query("category == 'GLOBAL_EXTREME'")["index"]
+                median = cems.loc[df["index"]].query(f"{field} > 0")[field].median()
+                screening.append(
+                    pd.concat(
+                        [
+                            pd.DataFrame(
+                                df.groupby("category").size().to_dict(),
+                                index=pd.MultiIndex.from_tuples([ts[0]]),
+                            ),
+                            pd.DataFrame(
+                                {
+                                    "MEAN_DEVIATION": [
+                                        (
+                                            cems.loc[global_extreme_id][field] / median
```

Comment:
Just for more compact viewing, can this mean deviation value be rounded to 1 decimal point?

Reply:
Done

Reply:
Here is the output now:
```
2024-03-29 10:40:22 [WARNING] oge.oge.validation:2026 Global extreme detected in CEMS time series
2024-03-29 10:40:22 [WARNING] oge.oge.validation:2027 
                                   gross_generation_mwh                fuel_consumed_mmbtu                   co2_mass_lb                ba_code
                                         GLOBAL_EXTREME MEAN_DEVIATION      GLOBAL_EXTREME MEAN_DEVIATION GLOBAL_EXTREME MEAN_DEVIATION        
plant_id_eia emissions_unit_id_epa                                                                                                             
315          3                                    350.0           12.0                 NaN            NaN            NaN            NaN    CISO
             4                                    297.0           12.1                 NaN            NaN            NaN            NaN    CISO
356          5                                     55.0           11.5                 NaN            NaN            NaN            NaN    CISO
             6                                    122.0           11.5                 NaN            NaN            NaN            NaN    CISO
377          5                                      NaN            NaN               113.0           10.6          157.0           10.7    LDWP
563          14B                                    NaN            NaN                 4.0           15.8            4.0           16.4    ISNE
673          S-3                                    NaN            NaN                 6.0           62.8            6.0           66.1    FMPP
874          5                                      NaN            NaN                44.0           37.9           44.0           37.7     PJM
1554         1                                      NaN            NaN                14.0           12.1           28.0           13.6     PJM
             4                                      NaN            NaN                12.0           14.2           12.0           19.2     PJM
1571         SMECO                                  NaN            NaN                 5.0           19.7            5.0           14.4     PJM
1702         A                                      NaN            NaN               156.0           30.6          154.0           30.2    MISO
             B                                      NaN            NaN               343.0           31.1          342.0           31.1    MISO
2049         CTA                                    NaN            NaN                 3.0           10.6            NaN            NaN    SOCO
             CTB                                    NaN            NaN                 3.0           10.6            NaN            NaN    SOCO
2081         11                                    17.0           13.5                 8.0           13.7            NaN            NaN    SWPP
             12                                     7.0           12.0                 1.0           15.0            NaN            NaN    SWPP
             13                                     6.0           14.7                 NaN            NaN            NaN            NaN    SWPP
             14                                    18.0           12.2                 6.0           12.3            NaN            NaN    SWPP
             15                                    15.0           13.2                 4.0           11.5            NaN            NaN    SWPP
             16                                    11.0           13.8                 5.0           12.1            NaN            NaN    SWPP
             17                                    25.0           14.8                10.0           11.8            NaN            NaN    SWPP
             18                                    25.0           14.3                10.0           13.0            NaN            NaN    SWPP
2499         CT01-1                                 NaN            NaN                 2.0           17.6            2.0           17.6    NYIS
             CT01-2                                 NaN            NaN                 1.0           10.2            1.0           10.2    NYIS
             CT01-3                                 NaN            NaN                 1.0           10.2            1.0           10.2    NYIS
             CT01-8                                 NaN            NaN                 3.0           16.1            3.0           15.8    NYIS
             CT02-1                                 NaN            NaN                 1.0           10.6            1.0           10.6    NYIS
             CT02-2                                 NaN            NaN                 1.0           12.4            1.0           12.3    NYIS
             CT02-3                                 NaN            NaN                 1.0           10.7            1.0           10.7    NYIS
             CT02-4                                 NaN            NaN                 1.0           10.5            1.0           10.5    NYIS
             CT02-5                                 NaN            NaN                 3.0           15.4            3.0           15.4    NYIS
             CT02-6                                 NaN            NaN                 1.0           12.0            1.0           11.9    NYIS
             CT02-7                                 NaN            NaN                 1.0           10.6            1.0           10.6    NYIS
             CT02-8                                 NaN            NaN                 3.0           16.8            3.0           16.8    NYIS
2632         1                                      NaN            NaN                 5.0           10.9            5.0           10.9    NYIS
3161         3                                      NaN            NaN                 NaN            NaN            2.0           10.3     PJM
3403         GCT2                                   2.0           10.5                 1.0           11.0            1.0           10.9     TVA
3576         BW2                                    NaN            NaN               139.0           14.1          139.0           14.0    ERCO
             BW3                                    NaN            NaN                88.0           16.1           90.0           16.1    ERCO
3809         3                                      NaN            NaN                94.0           23.0           87.0           14.4     PJM
3992         9                                      NaN            NaN                 5.0           10.4            5.0           10.4    MISO
4266         4                                    507.0           24.5                41.0           10.6           43.0           10.6    ERCO
             5                                    582.0           25.8               140.0           12.5          140.0           12.5    ERCO
6042         PMT1                                   NaN            NaN                 3.0           11.5            3.0           11.6     FPL
6074         1                                      8.0           10.1                 2.0           10.3            NaN            NaN    SWPP
             2                                      1.0           10.5                 1.0           12.7            NaN            NaN    SWPP
             3                                      1.0           10.0                 1.0           10.0            NaN            NaN    SWPP
             4                                      1.0           10.8                 1.0           10.1            NaN            NaN    SWPP
6085         16A                                    2.0           17.2                 NaN            NaN            NaN            NaN    MISO
6651         CT01                                   NaN            NaN                 6.0           26.9            NaN            NaN    MISO
6824         1                                     11.0           17.4                17.0           25.6            NaN            NaN    MISO
             2                                     12.0           16.6                17.0           25.2            NaN            NaN    MISO
7425         1                                      NaN            NaN                 9.0           17.1            NaN            NaN    MISO
10350        CTGB                                   NaN            NaN                 7.0           16.7            7.0           16.9    CISO
50732        ETBLR1                                 NaN            NaN                33.0           25.2            NaN            NaN     PJM
             ETBLR2                                 NaN            NaN                24.0           33.6            NaN            NaN     PJM
             ETBLR3                                 NaN            NaN                25.0           33.6            NaN            NaN     PJM
55096        BLR2                                   NaN            NaN                 4.0           14.0            NaN            NaN    MISO
55381        CT-005                                 NaN            NaN               119.0           13.1          119.0           13.0     PJM
55419        600                                    NaN            NaN                 NaN            NaN           22.0           28.4    MISO
             700                                    NaN            NaN                 NaN            NaN           29.0           33.4    MISO
             800                                    NaN            NaN                 NaN            NaN           28.0           34.0    MISO
```